src/config/lockfile.py

```diff
@@ -74,3 +106,113 @@ def write(self):
+            elif type is Asset.RESOURCE:
+                removed = self.lockfile_entries.resoucepacks.pop(entry.name)
+            elif type is Asset.TEXTURE:
+                removed = self.lockfile_entries.texturepacks.pop(entry.name)
+            else:
+                removed = self.lockfile_entries.shaderpacks.pop(entry.name)
+
+            return removed
+        return None
+
+    def update_entry(self, entry: LockfileEntry, entry_type: Asset):
+        """Updates an existing entry in the lockfile with a new LockfileEntry 
+        object.
+
+        If the object does not exist, adds it to the lockfile as a new entry.
+        """
+        if self.get_entry(entry, entry_type) is not None:
+            if type is Asset.MOD:
+                self.lockfile_entries.mods[entry.name] = entry
+            elif type is Asset.RESOURCE:
+                self.lockfile_entries.resoucepacks[entry.name] = entry
+            elif type is Asset.TEXTURE:
+                self.lockfile_entries.texturepacks[entry.name] = entry
+            else:
+                self.lockfile_entries.shaderpacks[entry.name] = entry
+        else:
+            self.add_entry(entry, entry_type)
+
+    def get_entry(self, entry: LockfileEntry, entry_type: Asset):
+        """Gets a specified entry from the lockfile object."""
+        if entry_type is Asset.MOD:
+            return self.lockfile_entries.mods.get(entry.name)
+        if entry_type is Asset.RESOURCE:
+            return self.lockfile_entries.resoucepacks.get(entry.name)
+        if entry_type is Asset.TEXTURE:
+            return self.lockfile_entries.texturepacks.get(entry.name)
+        return self.lockfile_entries.shaderpacks.get(entry.name)
+
+    def reindex_lockfile_entries(
+            self, lockfile_entries: LockfileEntries, f_new_key: Callable
+    ):
+        """Reindex lockfile entries across all types on a common new key.
+
+        Args:
+            lockfile_entries: the object containing all lockfile entry types
+            new_key: the new key to reindex lockfile entries by
+
+        Returns:
+            dict of reindexed lockfile entries, indexed by type
+        """
+
+        try:
+            reindexed_dicts = {
+                'mod_entries': reindex(
+                    lockfile_entries.mod_entries, f_new_key
+                ),
+                'resource_entries': reindex(
+                    lockfile_entries.resouce_entries, f_new_key
+                ),
+                'texture_entries': reindex(
+                    lockfile_entries.texture_entries, f_new_key
+                ),
+                'shader_entries': reindex(
+                    lockfile_entries.shader_entries, f_new_key
+                )
+            }
+        except Exception as error:
+            raise ClickException(
+                'Failed to reindex lockfile entries:' + f'{error}') from error
+        return reindexed_dicts
+
+
+class LockfileContextManager():
```

Comment:
These are probably separate enough to warrant their own files.

`lockfile_context_manager.py`